### ISSUE TITLE
Fails to load dependency graph when dependency uses Version Range

### ISSUE THREAD

**Issue by chrsmutti** [ckipp01/mill-github-dependency-graph#40]
First of all, thanks for the amazing work on this tool, really easy to use and it'll be an incredible selling point for me to pitch Mill and a more modern build process for some services at work!

I was testing this in an internal repository and had some issues with the dependency graph not loading, after some debugging (adding projects incrementally and reading the output of `generate`), I've pinpointed the issue to [Version Ranges](https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN402).

One of the dependencies of this project is `org.jongo:jongo:1.5.0`, with depends on `com.fasterxml.jackson.core:jackson-core:[2.7.0,2.12.3]`. When the manifest is generated with this type of version range, the Dependency Graph page fails to load, downgrading to `org.jongo:jongo:1.4.1` that does not include version ranges for Jackson works perfectly.

Github has some way of dealing with this, [if the project has a `pom.xml`](https://github.com/bguerout/jongo/blob/master/pom.xml#L76) it [shows the version range correctly](https://github.com/bguerout/jongo/network/dependencies), as seen here:
<img width="901" alt="image" src="https://user-images.githubusercontent.com/34144667/178996004-a4707617-dbd3-477c-8e75-e2dfc776246a.png">

But I did not find anything on the documentation of the Dependency Submission API.

**Comment by ckipp01** [comment]
Hey @chrsmutti, thanks for the report! Yes! I hit on the in `scala-cli` as well and it was super confusing. It turns out that the submission API doesn't support ranges since you have to have a valid [PURL](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst), which don't support ranges. That's totally fine because in reality, we don't want to display a range here, we actually want to display the resolved result. I'm not sure what version of `mill-github-dependency-graph` you're using, but this should be fixed in 0.1.0 that I released yesterday. If you're not on the latest version, could you check with that? I also plan on actually adding in https://github.com/package-url/packageurl-java to help with this, because there are some other oddities in PURLs.

Also, if you haven't seen I also created https://github.com/ckipp01/mill-dependency-submission which should be bringing in the latest version with this fix.

**Comment by chrsmutti** [comment]
I'm actually using the dependency submission action already:
<img width="1497" alt="image" src="https://user-images.githubusercontent.com/34144667/179009343-5eb14f11-7ce3-4542-b131-0c3379c1a311.png">

An excerpt of the generated manifest using `./millw --import ivy:io.chris-kipp::mill-github-dependency-graph::0.1.0 show io.kipp.mill.github.dependency.graph.Graph/generate`:

```
      "com.fasterxml.jackson.core:jackson-databind:[2.7.0,2.12.3]": {
        "metadata": {
          
        },
        "dependencies": [
          "com.fasterxml.jackson.core:jackson-annotations:2.12.3",
          "com.fasterxml.jackson.core:jackson-core:2.12.3"
        ],
        "relationship": "indirect",
        "package_url": "pkg:maven/com.fasterxml.jackson.core/jackson-databind@[2.7.0,2.12.3]"
      }
```

**Comment by ckipp01** [comment]
huh, that's very odd. I just tested this with:

```
import mill._, scalalib._

object minimal extends ScalaModule {
  def scalaVersion = "2.13.8"

  def ivyDeps = Agg(
    ivy"org.jongo:jongo:1.5.0"
  )
}
```
And then when running generate on it I see that jackson-databind is indeed resolved:

```
      "com.fasterxml.jackson.core:jackson-databind:2.12.3": {
        "metadata": {

        },
        "dependencies": [
          "com.fasterxml.jackson.core:jackson-annotations:2.12.3",
          "com.fasterxml.jackson.core:jackson-core:2.12.3"
        ],
        "relationship": "indirect",
        "package_url": "pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.12.3"
      },
```
Is this project public by chance that I could take a look at?

**Comment by chrsmutti** [comment]
Yeah, very odd indeed, I just tested a minimal version as well and had the same result as you. The project is internal 😞 

But I think I've reproduced it, it may be related to a module dep adding in some other version of the dependency, like so:

```scala
import mill._, scalalib._

object minimalDep extends ScalaModule {
  def scalaVersion = "2.13.8"

  def ivyDeps = Agg(
    ivy"com.fasterxml.jackson.core:jackson-core:2.13.3",
  )
}

object minimal extends ScalaModule {
  def moduleDeps = Seq(`minimalDep`)

  def scalaVersion = "2.13.8"

  def ivyDeps = Agg(
    ivy"org.jongo:jongo:1.5.0"
  )
}
```

This architecture one is closer to the original repo I was testing it, and results in:

```json
      "com.fasterxml.jackson.core:jackson-core:[2.7.0,2.12.3]": {
        "metadata": {

        },
        "dependencies": [

        ],
        "relationship": "indirect",
        "package_url": "pkg:maven/com.fasterxml.jackson.core/jackson-core@[2.7.0,2.12.3]"
      },
```

If `moduleDep` depends on `com.fasterxml.jackson.core:jackson-core` in the same range as `org.jongo:jongo`, the issue is also resolved. So it seems like an extreme edge case 😅

**Comment by lefou** [comment]
Question is, which version actually ends up in the classpath?

**Comment by ckipp01** [comment]
Ah perfect, thanks for the reproduction @chrsmutti. This is super helpful. I'll dig in and see what's going on because yea, this shouldn't be ending up here and I actually have no idea how it is since I'm grabbing the reconciled version every time. I'll dig in and report back.

**Comment by chrsmutti** [comment]
I'm really out of my league here as I don't have that much knowledge on dependency resolution, but coming from an user's perspective an error when the PURL is invalid seems like a good idea because the only information I have at the moment is this:
<img width="903" alt="image" src="https://user-images.githubusercontent.com/34144667/179044135-ea4e7397-aaf4-4607-bb7c-87febecd083c.png">

All that said, thank you very much for the quick answers!

_In this case in particular, I think I'm at fault as well, the upgraded dependency on the `minimalDep` is not on the original sbt project I was migrating, it should also depend on `jackson-*:2.12.3`, which would mean I wouldn't have caught this in the first place 😅_

**Comment by ckipp01** [comment]
> In this case in particular, I think I'm at fault as well, the upgraded dependency on the minimalDep is not on the original sbt project I was migrating, it should also depend on jackson-*:2.12.3, which would mean I wouldn't have caught this in the first place 😅

Ha, that's ok. These edge cases are helpful and hard to think of when first building something. So it's helpful see what people are hitting on.

**Comment by ckipp01** [comment]
pfff, they say PURL doesn't support ranges but you can still give one to their java library and end up with 
```
 "package_url": "pkg:maven/com.fasterxml.jackson.core/jackson-core@%5B2.7.0%2C2.12.3%5D"
```

Awesome.

**Comment by ckipp01** [comment]
Alright, everything is updated @chrsmutti. If you're using `v1` of the action you should be able to just try again and it will pull all the new stuff in. If that doesn't seem to, then just directly set the `plugin-version` in the GitHub action to use the latest `0.1.1`.

**Comment by chrsmutti** [comment]
It worked like a charm! Thank you so much for the quick fix

<img width="568" alt="image" src="https://user-images.githubusercontent.com/34144667/179071268-0eaf3fc1-d4de-4a6c-bf04-b2e3fdb19358.png">

**Comment by ckipp01** [comment]
Awesome, glad it works. I'm going to leave this open a bit because I'd still like to understand why that value is still a range in the tree.

**Comment by ckipp01** [comment]
Just to link these together, I created a discussion about this in https://github.com/coursier/coursier/discussions/2481.
